File: web/src/architecture/infrastructure/api/system-settings.ts
import { get, put, post } from '@/architecture/infrastructure/apiClient/request'

export type RegistrationMode = 'admin_only' | 'email_code' | 'debug_code'
export type EmailMode = 'smtp' | 'log'

export interface EmailSettings {
  mode: EmailMode
  host: string
  port: number
  username: string
  password?: string
  password_set: boolean
  from: string
  from_name: string
}

export interface SystemSettings {
  registration_mode: RegistrationMode
  email: EmailSettings
}

export interface SystemBackupConfig {
  enabled: boolean
  schedule_time: string
  endpoint: string
  region: string
  bucket: string
  prefix: string
  access_key_id: string
  secret_access_key?: string
  secret_access_key_set: boolean
  use_ssl: boolean
  force_path_style: boolean
  keep_local: number
  retention_days: number
}

export interface SystemBackupRecord {
  id: string
  triggered_by: string
  status: string
  started_at: string
  finished_at?: string
  archive_name?: string
  size_bytes?: number
  sha256?: string
  bucket?: string
  object_key?: string
  etag?: string
  error_message?: string
}

export interface SystemBackupOverview {
  config: SystemBackupConfig
  agent_available: boolean
  agent_last_seen_at?: string
  running: boolean
  records: SystemBackupRecord[]
}

export interface TLSCertificateInfo {
  subject: string
  issuer: string
  dns_names: string[]
  ip_addresses: string[]
  not_before: string
  not_after: string
  is_self_signed: boolean
}

export interface TLSSettings {
  mode: string
  base_url: string
  cert_file: string
  key_file: string
  cert_exists: boolean
  key_exists: boolean
  ready: boolean
  writable: boolean
  reload_supported: boolean
  certificate?: TLSCertificateInfo
  message?: string
}

export interface UpdateTLSCertificatePayload {
  certificate_pem: string
  private_key_pem: string
  reload: boolean
}

export interface AuthLoginProviderField {
  key: string
  label: string
  type: string
  required: boolean
  secret: boolean
  help?: string
  placeholder?: string
  value?: string
  value_set: boolean
}

export interface AuthLoginProviderInfo {
  code: string
  name: string
  description: string
  action: string
  enabled: boolean
  configured: boolean
  status: 'unconfigured' | 'disabled' | 'enabled' | string
  callback_path?: string
  docs_url?: string
  fields: AuthLoginProviderField[]
  updated_by?: string
  updated_at?: string
}

export interface ListAuthLoginProvidersResp {
  providers: AuthLoginProviderInfo[]
}

export interface LoginMethodInfo {
  provider: string
  label: string
  action: string
  description?: string
  authorize_path?: string
}

export interface ListLoginMethodsResp {
  methods: LoginMethodInfo[]
}

export interface LoginAnnouncement {
  enabled: boolean
  markdown: string
}

export interface LogArchiveResourceSummary {
  resource_path: string
  count: number
}

export interface LogArchiveBatch {
  id: number
  archive_key: string
  archive_type: string
  tenant_user: string
  app: string
  range_started_at: string
  range_ended_at: string
  record_count: number
  object_ref: string
  file_name: string
  file_size: number
  sha256: string
  status: 'exporting' | 'uploaded' | 'completed' | 'failed' | string
  summary_json?: { top_resource_paths?: LogArchiveResourceSummary[] }
  error_message?: string
  archived_at?: string
  deleted_at_source?: string
  created_at: string
}

export interface ListLogArchiveBatchesResp {
  list: LogArchiveBatch[]
  total: number
  retention_days: number
  cron_expr: string
  timezone: string
}

export interface SystemResourceComponent {
  key: string
  name: string
  pool_key: string
  used_bytes: number
  available: boolean
}

export interface SystemEnvironmentInfo {
  mode: 'development' | 'production' | string
  deployment: 'source' | 'aio' | 'container' | 'host' | string
  containerized: boolean
  container_engine: string
  container_remote: boolean
  storage_root_source: string
}

export interface SystemStoragePool {
  key: string
  name: string
  path?: string
  total_bytes: number
  used_bytes: number
  free_bytes: number
  used_percent: number
  primary: boolean
  available: boolean
}

export interface SystemResourceSnapshot {
  capacity_schema_version?: number
  collected_at: string
  hostname: string
  operating_system: string
  architecture: string
  cpu_cores: number
  cpu_used_percent: number
  cpu_available: boolean
  load_1: number
  load_5: number
  load_15: number
  load_available: boolean
  uptime_seconds: number
  memory_total_bytes: number
  memory_used_bytes: number
  memory_used_percent: number
  memory_available: boolean
  swap_total_bytes: number
  swap_used_bytes: number
  swap_used_percent: number
  network_rx_bytes: number
  network_tx_bytes: number
  network_available: boolean
  network_rx_bytes_per_second: number
  network_tx_bytes_per_second: number
  disk_read_bytes: number
  disk_write_bytes: number
  disk_io_available: boolean
  disk_read_bytes_per_second: number
  disk_write_bytes_per_second: number
  disk_mount: string
  disk_total_bytes: number
  disk_used_bytes: number
  disk_free_bytes: number
  disk_used_percent: number
  environment: SystemEnvironmentInfo
  storage_pools: SystemStoragePool[]
  components: SystemResourceComponent[]
  database_logical_bytes: number
  database_size_available: boolean
  database_inventory_complete: boolean
  databases: SystemDatabaseSize[]
  largest_databases: SystemDatabaseSize[]
}

export interface SystemResourceHistoryPoint {
  collected_at: string
  disk_used_bytes: number
  disk_used_percent: number
  memory_used_percent: number
  cpu_used_percent: number
  cpu_max_percent: number
  network_rx_bytes_per_second: number
  network_tx_bytes_per_second: number
  disk_read_bytes_per_second: number
  disk_write_bytes_per_second: number
  load_1: number
}

export interface SystemDatabaseSize {
  name: string
  kind: 'platform' | 'workspace' | string
  owner: string
  directory: string
  purpose: string
  status: 'active' | 'pending' | 'missing' | 'orphaned' | string
  used_bytes: number
}

export interface SystemPlatformMetrics {
  collected_at: string
  users_total: number
  users_active: number
  users_pending: number
  workspaces_total: number
  workspaces_enabled: number
  service_directories: number
  functions_total: number
  app_databases_total: number
  scheduled_tasks_total: number
  scheduled_tasks_active: number
  app_stats_available: boolean
  runtime_stats_available: boolean
  timer_stats_available: boolean
}

export interface SystemCollectionTaskStatus {
  key: 'runtime' | 'platform' | 'capacity' | string
  status: 'pending' | 'running' | 'success' | 'partial' | 'failed' | string
  last_started_at?: string
  last_succeeded_at?: string
  next_run_at?: string
  duration_millis: number
  error?: string
}

export interface StorageExpansionForecast {
  status: 'healthy' | 'warning' | 'critical' | string
  pool_key: string
  current_used_percent: number
  daily_growth_bytes: number
  target_percent: number
  days_to_target?: number
  message: string
}

export interface SystemCapacityDailyPoint {
  collected_at: string
  database_logical_bytes: number
  database_logical_delta: number
  database_logical_delta_available: boolean
  database_count: number
  database_count_delta: number
  database_count_delta_available: boolean
  platform_database_count: number
  workspace_database_count: number
  database_size_available: boolean
  database_count_available: boolean
}

export interface SystemResourceOverview {
  current: SystemResourceSnapshot
  history: SystemResourceHistoryPoint[]
  capacity_history: SystemCapacityDailyPoint[]
  history_hours: number
  sample_interval_minutes: number
  runtime_retention_days: number
  platform_retention_days: number
  capacity_retention_days: number
  platform_interval_hours: number
  capacity_interval_hours: number
  platform_schedule_local: string
  capacity_schedule_local: string
  capacity_collected_at: string
  forecast: StorageExpansionForecast
  platform: SystemPlatformMetrics
  collection_tasks: SystemCollectionTaskStatus[]
  runtime_interval_seconds: number
}

export interface SystemResourceSummary {
  current: SystemResourceSnapshot
  platform: SystemPlatformMetrics
  forecast: StorageExpansionForecast
  sample_interval_minutes: number
  runtime_retention_days: number
  runtime_interval_seconds: number
}

export interface SystemResourceTrends {
  history: SystemResourceHistoryPoint[]
  history_hours: number
  sample_interval_minutes: number
  runtime_retention_days: number
}

export interface SystemResourceStorage {
  collected_at: string
  environment: SystemEnvironmentInfo
  storage_pools: SystemStoragePool[]
  components: SystemResourceComponent[]
  forecast: StorageExpansionForecast
  capacity_retention_days: number
  capacity_schedule_local: string
}

export interface SystemResourceDatabaseList {
  items: SystemDatabaseSize[]
  total: number
  page: number
  page_size: number
  platform_count: number
  workspace_count: number
  database_logical_bytes: number
  database_size_available: boolean
  database_inventory_complete: boolean
  collected_at: string
  capacity_history: SystemCapacityDailyPoint[]
  capacity_retention_days: number
  capacity_schedule_local: string
}

export interface SystemResourceDiagnostics {
  collected_at: string
  environment: SystemEnvironmentInfo
  collection_tasks: SystemCollectionTaskStatus[]
  platform_retention_days: number
  capacity_retention_days: number
  platform_schedule_local: string
  capacity_schedule_local: string
  sample_interval_minutes: number
  runtime_retention_days: number
  runtime_interval_seconds: number
}

export interface SystemFunctionUsageItem {
  path: string
  name: string
  directory_path: string
  directory_name: string
  template_type: string
  total_calls: number
  period_calls: number
}

export interface SystemDirectoryUsageItem {
  path: string
  name: string
  function_count: number
  total_calls: number
  period_calls: number
}

export interface SystemUsageDailyPoint {
  date: string
  operations: number
  failed: number
}

export interface SystemUsageOverview {
  available: boolean
  collected_at: string
  period_days: number
  ranking_basis: 'period' | 'cumulative'
  operations_today: number
  operations_period: number
  failed_operations: number
  successful_calls: number
  top_directories: SystemDirectoryUsageItem[]
  top_functions: SystemFunctionUsageItem[]
  directory_total: number
  function_total: number
  ranking_page: number
  ranking_page_size: number
  daily_history: SystemUsageDailyPoint[]
  snapshot_schedule_local: string
}

export function getSystemSettings() {
  return get<SystemSettings>('/hr/api/v1/system/settings')
}

export function updateSystemSettings(data: SystemSettings) {
  return put<SystemSettings>('/hr/api/v1/system/settings', data)
}

export function testSystemEmail(to: string) {
  return post('/hr/api/v1/system/settings/email/test', { to })
}

export function getSystemBackupOverview() {
  return get<SystemBackupOverview>('/hr/api/v1/system/settings/backup')
}

export function updateSystemBackupConfig(data: SystemBackupConfig) {
  return put<SystemBackupOverview>('/hr/api/v1/system/settings/backup', data)
}

export function testSystemBackupS3(data: SystemBackupConfig) {
  return post('/hr/api/v1/system/settings/backup/test', data)
}

export function runSystemBackupNow() {
  return post<SystemBackupOverview>('/hr/api/v1/system/settings/backup/run')
}

export function getTLSSettings() {
  return get<TLSSettings>('/hr/api/v1/system/settings/tls')
}

export function updateTLSCertificate(data: UpdateTLSCertificatePayload) {
  return put<TLSSettings>('/hr/api/v1/system/settings/tls', data)
}

export function reloadTLSCertificate() {
  return post<TLSSettings>('/hr/api/v1/system/settings/tls/reload')
}

export function listAuthLoginProviders() {
  return get<ListAuthLoginProvidersResp>('/hr/api/v1/system/auth/providers')
}

export function getLoginAnnouncementConfig() {
  return get<LoginAnnouncement>('/hr/api/v1/system/auth/login-announcement')
}

export function updateLoginAnnouncementConfig(data: LoginAnnouncement) {
  return put<LoginAnnouncement>('/hr/api/v1/system/auth/login-announcement', data)
}

export function updateAuthLoginProviderConfig(code: string, config: Record<string, string>) {
  return put<AuthLoginProviderInfo>(`/hr/api/v1/system/auth/providers/${encodeURIComponent(code)}/config`, {
    config
  })
}

export function updateAuthLoginProviderEnabled(code: string, enabled: boolean) {
  return put<AuthLoginProviderInfo>(`/hr/api/v1/system/auth/providers/${encodeURIComponent(code)}/enabled`, {
    enabled
  })
}

export function listLoginMethods() {
  return get<ListLoginMethodsResp>('/hr/api/v1/auth/methods')
}

export function listLogArchiveBatches(page = 1, pageSize = 20) {
  return get<ListLogArchiveBatchesResp>('/workspace/api/v1/system/log_archives', { page, page_size: pageSize })
}

export function getSystemResourceOverview(hours = 24 * 7, includeHistory = true) {
  return get<SystemResourceOverview>('/hr/api/v1/system/settings/resources', {
    hours,
    include_history: includeHistory
  })
}

export function getSystemResourceSummary() {
  return get<SystemResourceSummary>('/hr/api/v1/system/settings/resources/summary')
}

export function getSystemResourceTrends(hours = 24 * 7) {
  return get<SystemResourceTrends>('/hr/api/v1/system/settings/resources/trends', { hours })
}

export function getSystemResourceStorage() {
  return get<SystemResourceStorage>('/hr/api/v1/system/settings/resources/storage')
}

export function getSystemResourceDatabases(params: { page?: number; page_size?: number; scope?: string; keyword?: string; include_history?: boolean } = {}) {
  return get<SystemResourceDatabaseList>('/hr/api/v1/system/settings/resources/databases', params)
}

export function getSystemResourceDiagnostics() {
  return get<SystemResourceDiagnostics>('/hr/api/v1/system/settings/resources/diagnostics')
}

export function getSystemResourceUsage(days = 7, page = 1, pageSize = 10) {
  return get<SystemUsageOverview>('/hr/api/v1/system/settings/resources/usage', { days, page, page_size: pageSize })
}

File: dto/system_settings.go
package dto

import "time"

type SystemBackupConfig struct {
	Enabled            bool   `json:"enabled"`
	ScheduleTime       string `json:"schedule_time"`
	Endpoint           string `json:"endpoint"`
	Region             string `json:"region"`
	Bucket             string `json:"bucket"`
	Prefix             string `json:"prefix"`
	AccessKeyID        string `json:"access_key_id"`
	SecretAccessKey    string `json:"secret_access_key,omitempty"`
	SecretAccessKeySet bool   `json:"secret_access_key_set"`
	UseSSL             bool   `json:"use_ssl"`
	ForcePathStyle     bool   `json:"force_path_style"`
	KeepLocal          int    `json:"keep_local"`
	RetentionDays      int    `json:"retention_days"`
}

type SystemBackupRecord struct {
	ID           string `json:"id"`
	TriggeredBy  string `json:"triggered_by"`
	Status       string `json:"status"`
	StartedAt    string `json:"started_at"`
	FinishedAt   string `json:"finished_at,omitempty"`
	ArchiveName  string `json:"archive_name,omitempty"`
	SizeBytes    int64  `json:"size_bytes,omitempty"`
	SHA256       string `json:"sha256,omitempty"`
	Bucket       string `json:"bucket,omitempty"`
	ObjectKey    string `json:"object_key,omitempty"`
	ETag         string `json:"etag,omitempty"`
	ErrorMessage string `json:"error_message,omitempty"`
}

type SystemBackupOverview struct {
	Config          SystemBackupConfig   `json:"config"`
	AgentAvailable  bool                 `json:"agent_available"`
	AgentLastSeenAt string               `json:"agent_last_seen_at,omitempty"`
	Running         bool                 `json:"running"`
	Records         []SystemBackupRecord `json:"records"`
}

type EmailSettings struct {
	Mode        string `json:"mode"`
	Host        string `json:"host"`
	Port        int    `json:"port"`
	Username    string `json:"username"`
	Password    string `json:"password,omitempty"`
	PasswordSet bool   `json:"password_set"`
	From        string `json:"from"`
	FromName    string `json:"from_name"`
}

type SystemSettingsResp struct {
	RegistrationMode string        `json:"registration_mode"`
	Email            EmailSettings `json:"email"`
}

type LoginAnnouncement struct {
	Enabled  bool   `json:"enabled"`
	Markdown string `json:"markdown"`
}

type UpdateLoginAnnouncementReq struct {
	Enabled  bool   `json:"enabled"`
	Markdown string `json:"markdown"`
}

type TLSCertificateInfo struct {
	Subject      string   `json:"subject"`
	Issuer       string   `json:"issuer"`
	DNSNames     []string `json:"dns_names"`
	IPAddresses  []string `json:"ip_addresses"`
	NotBefore    string   `json:"not_before"`
	NotAfter     string   `json:"not_after"`
	IsSelfSigned bool     `json:"is_self_signed"`
}

type TLSSettingsResp struct {
	Mode            string              `json:"mode"`
	BaseURL         string              `json:"base_url"`
	CertFile        string              `json:"cert_file"`
	KeyFile         string              `json:"key_file"`
	CertExists      bool                `json:"cert_exists"`
	KeyExists       bool                `json:"key_exists"`
	Ready           bool                `json:"ready"`
	Writable        bool                `json:"writable"`
	ReloadSupported bool                `json:"reload_supported"`
	Certificate     *TLSCertificateInfo `json:"certificate,omitempty"`
	Message         string              `json:"message,omitempty"`
}

type UpdateTLSCertificateReq struct {
	CertificatePEM string `json:"certificate_pem" binding:"required"`
	PrivateKeyPEM  string `json:"private_key_pem" binding:"required"`
	Reload         bool   `json:"reload"`
}

type UpdateSystemSettingsReq struct {
	RegistrationMode string        `json:"registration_mode" binding:"required,oneof=admin_only email_code debug_code"`
	Email            EmailSettings `json:"email"`
}

type TestEmailReq struct {
	To string `json:"to" binding:"required,email"`
}

type SystemResourceComponent struct {
	Key       string `json:"key"`
	Name      string `json:"name"`
	PoolKey   string `json:"pool_key"`
	UsedBytes uint64 `json:"used_bytes"`
	Available bool   `json:"available"`
}

type SystemEnvironmentInfo struct {
	Mode              string `json:"mode"`
	Deployment        string `json:"deployment"`
	Containerized     bool   `json:"containerized"`
	ContainerEngine   string `json:"container_engine"`
	ContainerRemote   bool   `json:"container_remote"`
	StorageRootSource string `json:"storage_root_source"`
}

type SystemStoragePool struct {
	Key         string  `json:"key"`
	Name        string  `json:"name"`
	Path        string  `json:"path,omitempty"`
	TotalBytes  uint64  `json:"total_bytes"`
	UsedBytes   uint64  `json:"used_bytes"`
	FreeBytes   uint64  `json:"free_bytes"`
	UsedPercent float64 `json:"used_percent"`
	Primary     bool    `json:"primary"`
	Available   bool    `json:"available"`
}

type SystemResourceSnapshot struct {
	CapacitySchemaVersion     int                       `json:"capacity_schema_version,omitempty"`
	CollectedAt               time.Time                 `json:"collected_at"`
	Hostname                  string                    `json:"hostname"`
	OperatingSystem           string                    `json:"operating_system"`
	Architecture              string                    `json:"architecture"`
	CPUCores                  int                       `json:"cpu_cores"`
	CPUUsedPercent            float64                   `json:"cpu_used_percent"`
	CPUAvailable              bool                      `json:"cpu_available"`
	Load1                     float64                   `json:"load_1"`
	Load5                     float64                   `json:"load_5"`
	Load15                    float64                   `json:"load_15"`
	LoadAvailable             bool                      `json:"load_available"`
	UptimeSeconds             uint64                    `json:"uptime_seconds"`
	MemoryTotalBytes          uint64                    `json:"memory_total_bytes"`
	MemoryUsedBytes           uint64                    `json:"memory_used_bytes"`
	MemoryUsedPercent         float64                   `json:"memory_used_percent"`
	MemoryAvailable           bool                      `json:"memory_available"`
	SwapTotalBytes            uint64                    `json:"swap_total_bytes"`
	SwapUsedBytes             uint64                    `json:"swap_used_bytes"`
	SwapUsedPercent           float64                   `json:"swap_used_percent"`
	NetworkRxBytes            uint64                    `json:"network_rx_bytes"`
	NetworkTxBytes            uint64                    `json:"network_tx_bytes"`
	NetworkAvailable          bool                      `json:"network_available"`
	NetworkRxBytesPS          float64                   `json:"network_rx_bytes_per_second"`
	NetworkTxBytesPS          float64                   `json:"network_tx_bytes_per_second"`
	DiskReadBytes             uint64                    `json:"disk_read_bytes"`
	DiskWriteBytes            uint64                    `json:"disk_write_bytes"`
	DiskIOAvailable           bool                      `json:"disk_io_available"`
	DiskReadBytesPS           float64                   `json:"disk_read_bytes_per_second"`
	DiskWriteBytesPS          float64                   `json:"disk_write_bytes_per_second"`
	DiskMount                 string                    `json:"disk_mount"`
	DiskTotalBytes            uint64                    `json:"disk_total_bytes"`
	DiskUsedBytes             uint64                    `json:"disk_used_bytes"`
	DiskFreeBytes             uint64                    `json:"disk_free_bytes"`
	DiskUsedPercent           float64                   `json:"disk_used_percent"`
	Environment               SystemEnvironmentInfo     `json:"environment"`
	StoragePools              []SystemStoragePool       `json:"storage_pools"`
	Components                []SystemResourceComponent `json:"components"`
	DatabaseLogicalBytes      uint64                    `json:"database_logical_bytes"`
	DatabaseSizeAvailable     bool                      `json:"database_size_available"`
	DatabaseInventoryComplete bool                      `json:"database_inventory_complete"`
	Databases                 []SystemDatabaseSize      `json:"databases"`
	// LargestDatabases is kept for snapshots created before the complete
	// database inventory was introduced. New collections leave it empty.
	LargestDatabases []SystemDatabaseSize `json:"largest_databases"`
}

type SystemResourceHistoryPoint struct {
	CollectedAt       time.Time `json:"collected_at"`
	DiskUsedBytes     uint64    `json:"disk_used_bytes"`
	DiskUsedPercent   float64   `json:"disk_used_percent"`
	MemoryUsedPercent float64   `json:"memory_used_percent"`
	CPUUsedPercent    float64   `json:"cpu_used_percent"`
	CPUMaxPercent     float64   `json:"cpu_max_percent"`
	NetworkRxBytesPS  float64   `json:"network_rx_bytes_per_second"`
	NetworkTxBytesPS  float64   `json:"network_tx_bytes_per_second"`
	DiskReadBytesPS   float64   `json:"disk_read_bytes_per_second"`
	DiskWriteBytesPS  float64   `json:"disk_write_bytes_per_second"`
	Load1             float64   `json:"load_1"`
}

type SystemDatabaseSize struct {
	Name      string `json:"name"`
	Kind      string `json:"kind"`
	Owner     string `json:"owner"`
	Directory string `json:"directory"`
	Purpose   string `json:"purpose"`
	Status    string `json:"status"`
	UsedBytes uint64 `json:"used_bytes"`
}

type SystemPlatformServiceStats struct {
	WorkspacesTotal      int64               `json:"workspaces_total"`
	WorkspacesEnabled    int64               `json:"workspaces_enabled"`
	ServiceDirectories   int64               `json:"service_directories"`
	FunctionsTotal       int64               `json:"functions_total"`
	AppDatabasesTotal    int64               `json:"app_databases_total"`
	ScheduledTasksTotal  int64               `json:"scheduled_tasks_total"`
	ScheduledTasksActive int64               `json:"scheduled_tasks_active"`
	Usage                SystemUsageSnapshot `json:"usage"`
}

// SystemFunctionUsageSnapshot is the cumulative successful invocation count
// already maintained by service_tree. Daily platform snapshots persist these
// compact counters so period deltas do not require request-level metric rows.
type SystemFunctionUsageSnapshot struct {
	Path          string `json:"path"`
	Name          string `json:"name"`
	DirectoryPath string `json:"directory_path"`
	DirectoryName string `json:"directory_name"`
	TemplateType  string `json:"template_type"`
	TotalCalls    int64  `json:"total_calls"`
}

type SystemUsageSnapshot struct {
	Available                  bool                          `json:"available"`
	CollectedAt                time.Time                     `json:"collected_at"`
	OperationDay               string                        `json:"operation_day"`
	OperationsToday            int64                         `json:"operations_today"`
	OperationsYesterday        int64                         `json:"operations_yesterday"`
	OperationsLast7Days        int64                         `json:"operations_last_7_days"`
	OperationsLast30Days       int64                         `json:"operations_last_30_days"`
	FailedOperationsToday      int64                         `json:"failed_operations_today"`
	FailedOperationsYesterday  int64                         `json:"failed_operations_yesterday"`
	FailedOperationsLast7Days  int64                         `json:"failed_operations_last_7_days"`
	FailedOperationsLast30Days int64                         `json:"failed_operations_last_30_days"`
	Functions                  []SystemFunctionUsageSnapshot `json:"functions"`
}

type SystemFunctionUsageItem struct {
	SystemFunctionUsageSnapshot
	PeriodCalls int64 `json:"period_calls"`
}

type SystemDirectoryUsageItem struct {
	Path          string `json:"path"`
	Name          string `json:"name"`
	FunctionCount int    `json:"function_count"`
	TotalCalls    int64  `json:"total_calls"`
	PeriodCalls   int64  `json:"period_calls"`
}

type SystemUsageDailyPoint struct {
	Date       string `json:"date"`
	Operations int64  `json:"operations"`
	Failed     int64  `json:"failed"`
}

type SystemUsageOverviewResp struct {
	Available             bool                       `json:"available"`
	CollectedAt           time.Time                  `json:"collected_at"`
	PeriodDays            int                        `json:"period_days"`
	RankingBasis          string                     `json:"ranking_basis"`
	OperationsToday       int64                      `json:"operations_today"`
	OperationsPeriod      int64                      `json:"operations_period"`
	FailedOperations      int64                      `json:"failed_operations"`
	SuccessfulCalls       int64                      `json:"successful_calls"`
	TopDirectories        []SystemDirectoryUsageItem `json:"top_directories"`
	TopFunctions          []SystemFunctionUsageItem  `json:"top_functions"`
	DirectoryTotal        int                        `json:"directory_total"`
	FunctionTotal         int                        `json:"function_total"`
	RankingPage           int                        `json:"ranking_page"`
	RankingPageSize       int                        `json:"ranking_page_size"`
	DailyHistory          []SystemUsageDailyPoint    `json:"daily_history"`
	SnapshotScheduleLocal string                     `json:"snapshot_schedule_local"`
}

type SystemDatabaseCapacityStats struct {
	Available  bool                 `json:"available"`
	TotalBytes uint64               `json:"total_bytes"`
	Databases  []SystemDatabaseSize `json:"databases"`
}

type SystemPlatformMetrics struct {
	CollectedAt           time.Time           `json:"collected_at"`
	UsersTotal            int64               `json:"users_total"`
	UsersActive           int64               `json:"users_active"`
	UsersPending          int64               `json:"users_pending"`
	WorkspacesTotal       int64               `json:"workspaces_total"`
	WorkspacesEnabled     int64               `json:"workspaces_enabled"`
	ServiceDirectories    int64               `json:"service_directories"`
	FunctionsTotal        int64               `json:"functions_total"`
	AppDatabasesTotal     int64               `json:"app_databases_total"`
	ScheduledTasksTotal   int64               `json:"scheduled_tasks_total"`
	ScheduledTasksActive  int64               `json:"scheduled_tasks_active"`
	AppStatsAvailable     bool                `json:"app_stats_available"`
	RuntimeStatsAvailable bool                `json:"runtime_stats_available"`
	TimerStatsAvailable   bool                `json:"timer_stats_available"`
	Usage                 SystemUsageSnapshot `json:"usage"`
}

type SystemCollectionTaskStatus struct {
	Key             string     `json:"key"`
	Status          string     `json:"status"`
	LastStartedAt   *time.Time `json:"last_started_at,omitempty"`
	LastSucceededAt *time.Time `json:"last_succeeded_at,omitempty"`
	NextRunAt       *time.Time `json:"next_run_at,omitempty"`
	DurationMillis  int64      `json:"duration_millis"`
	Error           string     `json:"error,omitempty"`
}

type StorageExpansionForecast struct {
	Status             string  `json:"status"`
	PoolKey            string  `json:"pool_key"`
	CurrentUsedPercent float64 `json:"current_used_percent"`
	DailyGrowthByte    float64 `json:"daily_growth_bytes"`
	TargetPercent      float64 `json:"target_percent"`
	DaysToTarget       *int    `json:"days_to_target,omitempty"`
	Message            string  `json:"message"`
}

type SystemCapacityDailyPoint struct {
	CollectedAt                   time.Time `json:"collected_at"`
	DatabaseLogicalBytes          uint64    `json:"database_logical_bytes"`
	DatabaseLogicalDelta          int64     `json:"database_logical_delta"`
	DatabaseLogicalDeltaAvailable bool      `json:"database_logical_delta_available"`
	DatabaseCount                 int       `json:"database_count"`
	DatabaseCountDelta            int       `json:"database_count_delta"`
	DatabaseCountDeltaAvailable   bool      `json:"database_count_delta_available"`
	PlatformDatabaseCount         int       `json:"platform_database_count"`
	WorkspaceDatabaseCount        int       `json:"workspace_database_count"`
	DatabaseSizeAvailable         bool      `json:"database_size_available"`
	DatabaseCountAvailable        bool      `json:"database_count_available"`
}

type SystemResourceOverviewResp struct {
	Current                SystemResourceSnapshot       `json:"current"`
	History                []SystemResourceHistoryPoint `json:"history"`
	CapacityHistory        []SystemCapacityDailyPoint   `json:"capacity_history"`
	HistoryHours           int                          `json:"history_hours"`
	SampleIntervalMinutes  int                          `json:"sample_interval_minutes"`
	RuntimeRetentionDays   int                          `json:"runtime_retention_days"`
	PlatformRetentionDays  int                          `json:"platform_retention_days"`
	CapacityRetentionDays  int                          `json:"capacity_retention_days"`
	PlatformIntervalHours  int                          `json:"platform_interval_hours"`
	CapacityIntervalHours  int                          `json:"capacity_interval_hours"`
	PlatformScheduleLocal  string                       `json:"platform_schedule_local"`
	CapacityScheduleLocal  string                       `json:"capacity_schedule_local"`
	CapacityCollectedAt    time.Time                    `json:"capacity_collected_at"`
	Forecast               StorageExpansionForecast     `json:"forecast"`
	Platform               SystemPlatformMetrics        `json:"platform"`
	CollectionTasks        []SystemCollectionTaskStatus `json:"collection_tasks"`
	RuntimeIntervalSeconds int                          `json:"runtime_interval_seconds"`
}

// SystemResourceSummaryResp is the small, frequently refreshed monitoring
// payload. Capacity inventories and historical series intentionally live in
// separate endpoints so a live dashboard does not retransmit daily data.
type SystemResourceSummaryResp struct {
	Current                SystemResourceSnapshot   `json:"current"`
	Platform               SystemPlatformMetrics    `json:"platform"`
	Forecast               StorageExpansionForecast `json:"forecast"`
	SampleIntervalMinutes  int                      `json:"sample_interval_minutes"`
	RuntimeRetentionDays   int                      `json:"runtime_retention_days"`
	RuntimeIntervalSeconds int                      `json:"runtime_interval_seconds"`
}

type SystemResourceTrendsResp struct {
	History               []SystemResourceHistoryPoint `json:"history"`
	HistoryHours          int                          `json:"history_hours"`
	SampleIntervalMinutes int                          `json:"sample_interval_minutes"`
	RuntimeRetentionDays  int                          `json:"runtime_retention_days"`
}

type SystemResourceStorageResp struct {
	CollectedAt           time.Time                 `json:"collected_at"`
	Environment           SystemEnvironmentInfo     `json:"environment"`
	StoragePools          []SystemStoragePool       `json:"storage_pools"`
	Components            []SystemResourceComponent `json:"components"`
	Forecast              StorageExpansionForecast  `json:"forecast"`
	CapacityRetentionDays int                       `json:"capacity_retention_days"`
	CapacityScheduleLocal string                    `json:"capacity_schedule_local"`
}

type SystemResourceDatabaseListResp struct {
	Items                     []SystemDatabaseSize       `json:"items"`
	Total                     int                        `json:"total"`
	Page                      int                        `json:"page"`
	PageSize                  int                        `json:"page_size"`
	PlatformCount             int                        `json:"platform_count"`
	WorkspaceCount            int                        `json:"workspace_count"`
	DatabaseLogicalBytes      uint64                     `json:"database_logical_bytes"`
	DatabaseSizeAvailable     bool                       `json:"database_size_available"`
	DatabaseInventoryComplete bool                       `json:"database_inventory_complete"`
	CollectedAt               time.Time                  `json:"collected_at"`
	CapacityHistory           []SystemCapacityDailyPoint `json:"capacity_history,omitempty"`
	CapacityRetentionDays     int                        `json:"capacity_retention_days"`
	CapacityScheduleLocal     string                     `json:"capacity_schedule_local"`
}

type SystemResourceDiagnosticsResp struct {
	CollectedAt            time.Time                    `json:"collected_at"`
	Environment            SystemEnvironmentInfo        `json:"environment"`
	CollectionTasks        []SystemCollectionTaskStatus `json:"collection_tasks"`
	PlatformRetentionDays  int                          `json:"platform_retention_days"`
	CapacityRetentionDays  int                          `json:"capacity_retention_days"`
	PlatformScheduleLocal  string                       `json:"platform_schedule_local"`
	CapacityScheduleLocal  string                       `json:"capacity_schedule_local"`
	SampleIntervalMinutes  int                          `json:"sample_interval_minutes"`
	RuntimeRetentionDays   int                          `json:"runtime_retention_days"`
	RuntimeIntervalSeconds int                          `json:"runtime_interval_seconds"`
}
